--- a/user_headers/include/linux/netfilter/xt_layer7.h	1969-12-31 16:00:00.000000000 -0800
+++ b/user_headers/include/linux/netfilter/xt_layer7.h	2014-06-05 17:32:38.858235992 -0700
@@ -0,0 +1,14 @@
+#ifndef _XT_LAYER7_H
+#define _XT_LAYER7_H
+
+#define MAX_PATTERN_LEN 8192
+#define MAX_PROTOCOL_LEN 256
+
+struct xt_layer7_info {
+    char protocol[MAX_PROTOCOL_LEN];
+    char pattern[MAX_PATTERN_LEN];
+    u_int8_t invert;
+    u_int8_t pkt;
+};
+
+#endif /* _XT_LAYER7_H */
